src/registration.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2021 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{anyhow, Error};
+use data_encoding::HEXLOWER;
+use hmac::{Hmac, Mac, NewMac};
+use log::debug;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use sha1::Sha1;
+
+type HmacSha1 = Hmac<Sha1>;
+
+#[derive(Debug, Serialize)]
+struct RegistrationPayload {
```

Comment:
As previously, please either hide the `struct` inside a function, to make it clear where it's supposed to be used, or document it. Here and the other `struct`s of this module.